File: fs/py_before_energy/675__cut-off-trees-for-golf-event/cut-off-trees-for-golf-event_benchmark_v2.py
# Time:  O(t * (logt + m * n)), t is the number of trees
# Space: O(t + m * n)

import collections
import heapq


class Solution(object):
    def cutOffTree(self, forest):
        """
        :type forest: List[List[int]]
        :rtype: int
        """
        def dot(p1, p2):
            return p1[0]*p2[0]+p1[1]*p2[1]

        def minStep(p1, p2):
            min_steps = abs(p1[0]-p2[0])+abs(p1[1]-p2[1])
            closer, detour = [p1], []
            lookup = set()
            while True:
                if not closer:  # cannot find a path in the closer expansions
                    if not detour:  # no other possible path
                        return -1
                    # try other possible paths in detour expansions with extra 2-step cost
                    min_steps += 2
                    closer, detour = detour, closer
                i, j = closer.pop()
                if (i, j) == p2:
                    return min_steps
                if (i, j) not in lookup:
                    lookup.add((i, j))
                    for I, J in (i+1, j), (i-1, j), (i, j+1), (i, j-1):
                        if 0 <= I < m and 0 <= J < n and forest[I][J] and (I, J) not in lookup:
                            is_closer = dot((I-i, J-j), (p2[0]-i, p2[1]-j)) > 0
                            (closer if is_closer else detour).append((I, J))
            return min_steps

        m, n = len(forest), len(forest[0])
        min_heap = []
        for i in xrange(m):
            for j in xrange(n):
                if forest[i][j] > 1:
                    heapq.heappush(min_heap, (forest[i][j], (i, j)))

        start = (0, 0)
        result = 0
        while min_heap:
            tree = heapq.heappop(min_heap)
            step = minStep(start, tree[1])
            if step < 0:
                return -1
            result += step
            start = tree[1]
        return result


# Time:  O(t * (logt + m * n)), t is the number of trees
# Space: O(t + m * n)
class Solution_TLE(object):
    def cutOffTree(self, forest):
        """
        :type forest: List[List[int]]
        :rtype: int
        """
        def minStep(p1, p2):
            min_steps = 0
            lookup = {p1}
            q = collections.deque([p1])
            while q:
                size = len(q)
                for _ in xrange(size):
                    (i, j) = q.popleft()
                    if (i, j) == p2:
                        return min_steps
                    for i, j in (i+1, j), (i-1, j), (i, j+1), (i, j-1):
                        if not (0 <= i < m and 0 <= j < n and forest[i][j] and (i, j) not in lookup):
                            continue
                        q.append((i, j))
                        lookup.add((i, j))
                min_steps += 1
            return -1

        m, n = len(forest), len(forest[0])
        min_heap = []
        for i in xrange(m):
            for j in xrange(n):
                if forest[i][j] > 1:
                    heapq.heappush(min_heap, (forest[i][j], (i, j)))

        start = (0, 0)
        result = 0
        while min_heap:
            tree = heapq.heappop(min_heap)
            step = minStep(start, tree[1])
            if step < 0:
                return -1
            result += step
            start = tree[1]
        return result


if __name__ == "__main__":
    tests = [
        [[1]],
        [[1, 2]],
        [[1, 0], [3, 2]],
        [[1, 2, 3], [0, 0, 0], [7, 6, 5]],
        [[1, 2, 3, 4], [0, 0, 5, 0], [8, 7, 6, 9], [0, 0, 0, 10]],
        [[1, 1, 1, 1, 1], [0, 0, 0, 0, 1], [1, 2, 3, 4, 5], [1, 0, 0, 0, 0], [6, 7, 8, 9, 10]],
        [[1, 1], [1, 1]],
        [[1, 3, 5], [2, 8, 4], [7, 6, 9]],
        [[1, 2, 3, 4, 5, 6], [0, 0, 0, 0, 0, 7], [12, 11, 10, 9, 8, 14], [13, 0, 0, 0, 0, 15], [16, 17, 18, 19, 20, 21], [0, 0, 0, 0, 0, 22]],
        [[1, 0, 2, 0, 3], [4, 5, 6, 0, 7], [0, 0, 8, 9, 10]]
    ]

    sol = Solution()
    sink = 0

    iterations = 1000
    for _ in xrange(iterations):
        #checksum = 0
        for forest in tests:
            r= sol.cutOffTree(forest)
        #sink = checksum

File: fs/cpp_before_energy/675__cut-off-trees-for-golf-event/cut-off-trees-for-golf-event_benchmark_v2.cpp
#include <iostream>
template <class T>
static inline void DoNotOptimize(const T& value) {
    asm volatile("" : : "r,m"(value) : "memory");
}

#include <vector>
#include <tuple>
#include <algorithm>
#include <utility>
#include <cstdlib>

using namespace std;

// Time:  O(t * (logt + m * n)), t is the number of trees
// Space: O(t + m * n)
class Solution {
public:
    int cutOffTree(vector<vector<int>>& forest) {
        int m = forest.size();
        if (m == 0) return -1;
        int n = forest[0].size();
        if (n == 0) return -1;
        if (forest[0][0] == 0) return -1;

        auto dot = [](int x1, int y1, int x2, int y2) -> int {
            return x1 * x2 + y1 * y2;
        };

        auto minStep = [&](pair<int, int> p1, pair<int, int> p2) -> int {
            int min_steps = std::abs(p1.first - p2.first) + std::abs(p1.second - p2.second);
            vector<pair<int, int>> closer;
            vector<pair<int, int>> detour;
            closer.push_back(p1);
            vector<vector<bool>> visited(m, vector<bool>(n, false));

            while (true) {
                if (closer.empty()) {
                    if (detour.empty()) {
                        return -1;
                    }
                    min_steps += 2;
                    std::swap(closer, detour);
                }
                auto cur = closer.back();
                closer.pop_back();
                int i = cur.first, j = cur.second;
                if (i == p2.first && j == p2.second) {
                    return min_steps;
                }
                if (!visited[i][j]) {
                    visited[i][j] = true;
                    const int dirs[4][2] = {{1,0},{-1,0},{0,1},{0,-1}};
                    for (const auto& d : dirs) {
                        int I = i + d[0], J = j + d[1];
                        if (0 <= I && I < m && 0 <= J && J < n && forest[I][J] > 0 && !visited[I][J]) {
                            int move_x = I - i, move_y = J - j;
                            int to_target_x = p2.first - i, to_target_y = p2.second - j;
                            bool is_closer = dot(move_x, move_y, to_target_x, to_target_y) > 0;
                            if (is_closer) {
                                closer.emplace_back(I, J);
                            } else {
                                detour.emplace_back(I, J);
                            }
                        }
                    }
                }
            }
            // Unreachable
            // return min_steps;
        };

        vector<tuple<int, int, int>> trees;
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (forest[i][j] > 1) {
                    trees.emplace_back(forest[i][j], i, j);
                }
            }
        }
        sort(trees.begin(), trees.end());

        pair<int, int> start = {0, 0};
        int result = 0;
        for (const auto& t : trees) {
            int ti = std::get<1>(t);
            int tj = std::get<2>(t);
            int step = minStep(start, {ti, tj});
            if (step < 0) {
                return -1;
            }
            result += step;
            start = {ti, tj};
        }
        return result;
    }
};

int main() {
    vector<vector<vector<int>>> tests = {
        {{1}},
        {{1,2}},
        {{1,0},{3,2}},
        {{1,2,3},{0,0,0},{7,6,5}},
        {{1,2,3,4},{0,0,5,0},{8,7,6,9},{0,0,0,10}},
        {{1,1,1,1,1},{0,0,0,0,1},{1,2,3,4,5},{1,0,0,0,0},{6,7,8,9,10}},
        {{1,1},{1,1}},
        {{1,3,5},{2,8,4},{7,6,9}},
        {{1,2,3,4,5,6},{0,0,0,0,0,7},{12,11,10,9,8,14},{13,0,0,0,0,15},{16,17,18,19,20,21},{0,0,0,0,0,22}},
        {{1,0,2,0,3},{4,5,6,0,7},{0,0,8,9,10}}
    };

    Solution sol;
    volatile int sink = 0;

    const int iterations = 1000;
    for (int iter = 0; iter < iterations; ++iter) {
        //int checksum = 0;
        for (auto& forest : tests) {
            int r = sol.cutOffTree(forest);
            DoNotOptimize(r);
        }
        //sink = checksum;
    }
    return 0;
}